projects/fal/src/fal/api.py

```diff
@@ -873,10 +894,35 @@ def openapi(self) -> dict[str, Any]:
         return self._build_app().openapi()
 
     def serve(self) -> None:
-        import uvicorn
+        import asyncio
+
+        from starlette_exporter import handle_metrics
+        from uvicorn import Config
 
         app = self._build_app()
-        uvicorn.run(app, host="0.0.0.0", port=8080)
+        server = Server(config=Config(app, host="0.0.0.0", port=8080))
+        metrics_app = FastAPI()
+        metrics_app.add_route("/metrics", handle_metrics)
+        metrics_server = Server(config=Config(metrics_app, host="0.0.0.0", port=9090))
+
+        async def _serve() -> None:
+            event = asyncio.Event()
+            # TODO(squat): handle shutdowns gracefully.
+            # You cannot add signal handlers to any loop if you're not
+            # on the main thread.
+            # How can we detect that we are being shut down and stop the
+            # uvicorn servers gracefully?
+            # loop = asyncio.get_running_loop()
+            # loop.add_signal_handler(signal.SIGINT, event.set)
+            # loop.add_signal_handler(signal.SIGTERM, event.set)
```

Comment:
Have you considered running them in separate threads? Also interesting if there are implications to running two servers in the same event loop (basically same thread?).

Reply:
we run our two servers in the same event loop for all of our isolate cloud instances. We could run them in different threads, but that doesn't really fix this issue commented out here at all. We have to fix signal handling in the isolate agent to signal the thread in this fal app using a threading event, then this fal app can decide what to do with that event, e.g. set an asyncio event for async servers somehow call `stop` on the uvicorn workers. In short, threading vs async doesn't fix this graceful shutdown at all. I prefer to use async for our multiple servers. Some of our internal fal isolate servers have >3 servers running in the same event loop: grpc, http, metrics, etc. Eventually I'd like to define our async-friendly uvicorn server class here in fal-ai/fal and import it in our internal tooling.

Reply:
Oh, didn't know that we have an additional issue with the agent. Not a blocker for me, we can fix this when we need to.